File: app/src/main/java/com/example/android/sunshine/data/database/SunshineDatabase.java
package com.example.android.sunshine.data.database;

import android.arch.persistence.room.Database;
import android.arch.persistence.room.Room;
import android.arch.persistence.room.RoomDatabase;
import android.arch.persistence.room.TypeConverters;
import android.content.Context;
import android.util.Log;

@Database(entities = {WeatherEntry.class}, version = 1)
@TypeConverters(DateConverter.class)
public abstract class SunshineDatabase extends RoomDatabase {

    private static final String LOG_TAG = SunshineDatabase.class.getSimpleName();
    private static final String DATABASE_NAME = "weather";

    // For Singleton instantiation
    private static final Object LOCK = new Object();
    private static SunshineDatabase sInstance;

    public static SunshineDatabase getInstance(Context context) {
        Log.d(LOG_TAG, "Getting the database");
        if (sInstance == null) {
            synchronized (LOCK) {
                sInstance = Room.databaseBuilder(context.getApplicationContext(),
                        SunshineDatabase.class, SunshineDatabase.DATABASE_NAME).build();
                Log.d(LOG_TAG, "Made new database");
            }
        }
        return sInstance;
    }

    // The associated DAOs for the database
    public abstract WeatherDao weatherDao();
}

File: app/src/main/java/com/example/android/sunshine/data/database/SunshineDatabase.kt
package com.example.android.sunshine.data.database

import android.arch.persistence.room.Database
import android.arch.persistence.room.Room
import android.arch.persistence.room.RoomDatabase
import android.arch.persistence.room.TypeConverters
import android.content.Context

@Database(entities = arrayOf(WeatherEntry::class), version = 1)
@TypeConverters(DateConverter::class)
abstract class SunshineDatabase : RoomDatabase() {
    abstract fun weatherDao(): WeatherDao

    companion object {
        var instance: SunshineDatabase? = null
        const val DATABASE_NAME = "weather"

        fun getSunshineDatabase(context: Context): SunshineDatabase? {
            if (instance == null) {
                synchronized(SunshineDatabase::class) {
                    instance = Room.databaseBuilder(context.applicationContext,
                            SunshineDatabase::class.java, DATABASE_NAME)
                            .build()
                }
            }

            return instance
        }
    }
}